src/Serilog/Data/LogEventPropertyValueRewriter.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright 2016 Serilog Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using Serilog.Events;
+
+namespace Serilog.Data
+{
+    /// <summary>
+    /// A base class for visitors that rewrite the value with modifications. For example, implementations
+    /// might remove all structure properties with a certain name, apply size/length limits, or convert scalar properties of
+    /// one type into scalar properties of another.
+    /// </summary>
+    /// <typeparam name="TState"></typeparam>
+    public abstract class LogEventPropertyValueRewriter<TState> : LogEventPropertyValueVisitor<TState, LogEventPropertyValue>
+    {
+        /// <summary>
+        /// Visit a <see cref="ScalarValue"/> value.
+        /// </summary>
+        /// <param name="state">Operation state.</param>
+        /// <param name="scalar">The value to visit.</param>
+        /// <returns>The result of visiting <paramref name="scalar"/>.</returns>
+        protected override LogEventPropertyValue VisitScalarValue(TState state, ScalarValue scalar)
+        {
+            if (scalar == null) throw new ArgumentNullException(nameof(scalar));
+            return scalar;
+        }
+
+        /// <summary>
+        /// Visit a <see cref="SequenceValue"/> value.
+        /// </summary>
+        /// <param name="state">Operation state.</param>
+        /// <param name="sequence">The value to visit.</param>
+        /// <returns>The result of visiting <paramref name="sequence"/>.</returns>
+        protected override LogEventPropertyValue VisitSequenceValue(TState state, SequenceValue sequence)
+        {
+            if (sequence == null) throw new ArgumentNullException(nameof(sequence));
+
+            var contents = new LogEventPropertyValue[sequence.Elements.Count];
```

Comment:
If modification is unlikey, it'd be optimal to allocate this lazily and re-visit **iff** the `Visit()` call on an element returns a different reference to what was passed in.

As modification is probably the exception, rather than the rule, should we switch? Worst-case performance would be somewhat worse, but also consider:

 * If a rewriter _does_ change a lot of data, we'll probably trigger the lazy allocation on the very first element and then fall back to equivalent performance
 * The non-allocating work required to determine that allocation is necessary, is probably significantly quicker than any of the subsequent allocation/copying

(Same case for structures/dictionaries).